### ISSUE TITLE
[API Proposal]: AssemblyLoadContext.MixedAssemblyLoadContextFactory

### ISSUE THREAD

**Issue by szilvaa-adsk** [dotnet/runtime#62754]
### Background and motivation

The motivation for this proposal is #61105. .net host applications that use C++/CLI assemblies (aka mixed assemblies) will have a challenging time accommodating the new behavior dictated by .net6. This proposal allows these applications to customize the default behavior.

### API Proposal

```C#
namespace System.Runtime.Loader
{
    public class AssemblyLoadContext
    {
        public static Func<string, AssemblyLoadContext>? MixedAssemblyLoadContextFactory {get;set;}
    }
}
```
The implementation of this API is provided [here](https://github.com/dotnet/runtime/compare/main...szilvaa-adsk:main)

### API Usage

```C#
// Return default context to approximate .net 4.x behavior
AssemblyLoadContext.MixedAssemblyLoadContextFactory = (path) => AssemblyLoadContext.Default;
```

### Alternative Designs

An alternative design is to modify ijwhost.dll as implemented [here](https://github.com/dotnet/runtime/compare/release/6.0...szilvaa-adsk:release/6.0). This has the advantage that ijwhost.dll is not part of the runtime (distributed by the host app) but it is less flexible than what can be achieved using the proposed new API.

**Comment by vitek-karas** [comment]
I think there are two sides to this problem and I'm personally not sure yet which side should be the one to "win".
* The ability for the host to dictate where C++/CLI assemblies are loaded into (this proposal is one way to add such ability)
* The ability for the C++/CLI assembly to request where to be loaded

What already exists:
* If the C++/CLI assembly is loaded from managed code it follows existing ALC rules and will be loaded into the same ACL just like any other assembly. In this case the host has full control over it and the assembly has no say in it (just like any other assembly)
* Native hosting loading assembly into the process - currently this is almost always forced into a secondary (isolated) ALC. There's a way to load into default but it is currently only usable in special cases. There are plans to add the ability to specify the ALC in this case - see [`get_function_pointer_fn`](https://github.com/dotnet/runtime/blob/main/docs/design/features/native-hosting.md#calling-managed-function-net-5-and-above) which already has `load_context` parameter, and the plan is to add a new `load_assembly_fn` which would have the ability to specify the ALC to load into (as a counterpart to already existing `load_assembly_and_get_function_pointer_fn`). The ultimate goal is that the native hosting has full control over the load - basically exact opposite of the managed assembly load. Note that in this case the assembly also has no direct say in it - if the native hosting code is not coupled with the assembly.

C++/CLI is special in that it combines both of the above - it can be loaded as managed assembly or through native hosting (`ijwhost`). And to make things "worse" it inherently includes the native hosting path which is coupled with the assembly. There are just no ways to influence the native hosting path (unlike the managed loading which is fully controlled by the managed host).

COM components are very similar to C++/CLI. They are almost always loaded through native hosting (through COM itself) and currently offer no way to influence where they get loaded into (always into isolated ALC).

I think the question is which functionality is more important to have (I could see us having both, but I expect one to be much more useful over the other):
* Ability to override where assemblies loaded via native entry points get loaded into (would apply to both COM and C++/CLI and potentially even to direct native hosting in some cases I assume). This would likely be solved by a new API - something along the lines of this proposal.
* Ability for a component to specify where it wants to be loaded (would apply to COM and C++/CLI). I would assume this would be solved by adding a new configuration option to the `.runtimeconfig.json` of the component.

I just don't know which one of the two is more important to have. Currently I'm leaning a little to the second one where the component has the ability to ask to be loaded into default ALC:
* Some components know this about themselves as a simple compatibility thing - they rely on functionality which is broken for whatever reason when loaded into secondary ALC.
* Processes which don't have a managed host, where the managed component is treated as if it were native - this happens quite a bite for COM and can happen for C++/CLI as well. In this case the ability to override the behavior from the host is not very useful, as there's no code to put it. Especially if the component doesn't have a way to affect the host in any meaningful way (very common for COM).

In both cases having the ability to specify this on the component itself would solve the problem. The ability to override this from the host is only useful if the host is already running managed code and it knows that it will run into this problem.

/cc @AaronRobinsonMSFT @agocke

**Comment by AaronRobinsonMSFT** [comment]
> Ability for a component to specify where it wants to be loaded (would apply to COM and C++/CLI). I would assume this would be solved by adding a new configuration option to the .runtimeconfig.json of the component.

Definitely could be done this way. A counterpoint to this approach is the `.runtimeconfig.json` is for runtime configuration not really about the assembly &ndash; which I'd argue is simply an organization abstraction. An alternative approach might be to have an assembly level attribute that indicates where the assembly should be loaded. However, the complexity here is then we need to read the metadata prior to loading the assembly &ndash; ugh. This means we are likely back to one of the json files, which I dislike.

I strongly agree with @vitek-karas's statement from https://github.com/dotnet/runtime/issues/61105#issuecomment-958096989:

> The magic of fusion in .NET Framework worked reasonably well for simple cases, but more complex cases caused all kinds of trouble (probably the most known is binding redirect hell). .NET Core was intentionally designed to have a much simpler assembly loader, where most things are specified explicitly. This can be either by the SDK (.deps.json) or by the code (ALCs, and the various loader extension points).

To me, this means callbacks/events should be an absolute last resort since they permit creation of inscrutable assembly loading scenarios that will serve only to make for fragile solutions. I prefer a declarative approach that can be audited easily.

**Comment by szilvaa** [comment]
> If the C++/CLI assembly is loaded from managed code it follows existing ALC rules and will be loaded into the same ACL just like any other assembly. In this case the host has full control over it and the assembly has no say in it (just like any other assembly)

This suggests that "native loads" should be treated identically. Anything else is just an extra mental load on the client (me). In other words, the host should determine the ALC.

>To me, this means callbacks/events should be an absolute last resort since they permit creation of inscrutable assembly loading scenarios that will serve only to make for fragile solutions. I prefer a declarative approach that can be audited easily.

Again, when an assembly is loaded by managed code this is already a possibility. "Native load" shouldn't introduce new safeguards. It does not solve the problem and it adds to my learning curve.

**Comment by AaronRobinsonMSFT** [comment]
> Again, when an assembly is loaded by managed code this is already a possibility. "Native load" shouldn't introduce new safeguards.

I think this is a narrow way of looking at the problem. Managed and native assembly load are fundamentally different and therefore have different constraints and considerations. For the native load, when is the above callback going to be subscribed to? If this is the first time the runtime is being loaded, no managed code has run yet and therefore there is no one to set the proposed callback. If the runtime is already loaded then perhaps that is possible, but that is a narrower scenario. It could be that the runtime is already loaded scenario is the only interesting one but given the last few releases worth of feedback in this area, I don't think that is the case.

**Comment by szilvaa-adsk** [comment]
This callback is about *customization* by a host that cares. If the host does not care then they get the current behavior. If the host cares then it can load managed code and hook the callback prior to running code that may cause a "native load". This is what I do in my host app.

>It could be that the runtime is already loaded scenario is the only interesting one but given the last few releases worth of feedback in this area, I don't think that is the case.

I have no data on how far we (all users) are in migrating from .net framework to .net (core). The only data I have is the data from the company I work for that makes engineering apps. For complex desktop apps like ours .net6 is the first time when we are even attempting migration.

**Comment by vitek-karas** [comment]
I would prefer a solution which allows the piece of the system which causes the problem to also fix the problem. If the component being loaded doesn't work in a secondary ALC, it should state that about itself.

The host customization is basically a way to patch problems of components which would not do this. I could see value in both approaches, but currently it feels like:
* The ability of a component to declare its needs is important - since it applies (and fixes the problem) for all hosts, regardless if they "care" or not.
* The ability to override this behavior in the host only applies to hosts which already run managed code and are aware of the issues (or cause them).

I guess what I'm trying to say is that just adding the ability for host to customize the behavior doesn't feel like actually addressing the problem. I would prefer we do either only the first one (per-component) or both.

Note that for some hosts there is a workaround - if the host knows the C++/CLI assemblies which are going to be used, it can preload them via the managed APIs which allow specifying target ALC already. From a performance perspective doing this should not be too bad, CLR is very lazy and will do very little with the assembly when it's only loaded (and no code runs in it).

**Comment by szilvaa-adsk** [comment]
>I would prefer a solution which allows the piece of the system which causes the problem to also fix the problem. If the component being loaded doesn't work in a secondary ALC, it should state that about itself.

I think this would work for us but I don't know how to implement this. It also has the disadvantage that we have to implement the solution in dozens of places (each component) instead of one place.

I also don't think that it is the component's fault that it doesn't work when loaded in a secondary ALC. What we have is this:
Component1 and Component2 depend on Component0 and Component1 depends on Component2. Components may be mixed or pure. All components are independently loadable by the user and only one copy of each should exist in the process. This system may be fragile, but it works today with .net framework and has worked for 18 years. Secondary ALCs break this system. This is a host-specific problem suggesting a host-specific solution.

> Note that for some hosts there is a workaround - if the host knows the C++/CLI assemblies which are going to be used, it can preload them via the managed APIs which allow specifying target ALC already

Yes. In fact, I pursued this workaround as far as I could. Unfortunately, it turns out that the host does not know in advance all the mixed mode assemblies that it may encounter.

**Comment by szilvaa-adsk** [comment]
@vitek-karas How can we move this forward? Can I open a PR with the implementation [proposed](https://github.com/dotnet/runtime/compare/main...szilvaa-adsk:main)?

**Comment by vitek-karas** [comment]
Typically for this kind of change we would write down the proposal as a design doc. The key points in there are not necessarily the exact technical solution but what problem is it solving, what other approaches were considered and explanations of "why" + the reasonably detailed technical design of the change. Once that is done, if there's a new API proposed it will need to go through normal API review process. And only then we would start the real implementation.

In this case I really believe that we should not go with host-only solution. In general I think components should be self-descriptive and we need to provide features for them to do so. The host solution proposed here is in my mind secondary in that it would be typically used to fix the problems for components which are not self-descriptive enough.

For example in your case:
> All components are independently loadable by the user and only one copy of each should exist in the process.

I can definitely see cases where the "one per process" is a requirement of the component, not the host per se. ALCs don't completely break this, the host could decide to load all these components into one secondary ALC. But such behaviors would be outside of the scope of this.

I know this sounds like a lot of process and such... but if we accept the change, we will have to live with it for a very long time and maintain it. So it needs to have a good enough reason behind it - the cost is non-trivial.

**Comment by szilvaa-adsk** [comment]
I see. I think we disagree about what problem we are solving. I see this as a limited application migration (.net4 to .net6) issue. I think you view this as a broader, generic problem of component configuration. Let me try to explain my point of view one more time to see if I can convince you.

At this point, C++/CLI and mixed assemblies appear to be an evolutionary dead-end. The only people who have these assets and, therefore, this problem are the ones who come from .net4 or earlier. People are encouraged to use other forms of interop that are supported on non-Windows platforms. This is important because this shows that (1) the problem impacts an audience who come to this with expectations established by their historic usage (2) the problem has limited and shrinking scope.

In these earlier releases, loading a mixed assembly with `Assembly::Load` vs. `LoadLibrary` resulted in the same runtime behavior of the loaded code.  .Net6 changes this. Mixed assemblies now behave radically differently depending on whether they are loaded via managed vs. unmanaged entry point. The new behavior may be better, but it will break some percentage of apps being migrated. These apps will want to restore the "old" behavior for the purpose of migration before they rearchitect to take advantage of better isolation that .net6 offers. 

Finally, we come to the question of who should be in control of the ACL by which a mixed assembly is loaded when it is loaded via unmanaged entry point.  Should it be the host (caller) or the component (callee)? It seems to me that it should be the caller since the caller is in control of the ACL when the component is loaded via managed entry point.  Why should there be a difference between managed and unmanaged load?

**Comment by vitek-karas** [comment]
Thanks for the explanation. The scenario you describe definitely makes sense. My view was a bit broader as this issue applies to not just C++/CLI but also COM. And while I do agree users will hopefully move away from C++/CLI, COM support is here to stay for a while. For this reason I would like to solve this issue for both cases.

I think there are basically two major cases:
* The host has managed code which "acts" like the host - meaning it has sort of an "ownership" of the process. In this case giving some level of control to it makes sense. And so adding an API like you propose would probably work.
* The host is a native app, which doesn't really know that it's loading managed code. This is pretty common for COM, but I've seen it with C++/CLI as well. In this case there's really nobody to "give control" process wide. And thus the components themselves need to be able to describe their needs.

But maybe we don't need to solve both cases at once (they have different technical solutions anyway).

@AaronRobinsonMSFT @elinor-fung What do you think about giving managed hosts a hook into the native hosting process? Basically a way to hook into the process of loading assembly from native hosting.
I would probably change the hook method to look something like (ignore the naming please...):
```C#
Assembly LoadAssemblyForNativeHosting(string path) ...
```

The native host has no need for the ALC itself - it really needs the assembly. We could make it that if this returns null we fallback to the current behavior (to give the hook the option to not mess up with assemblies it knows nothing about for example).

We would also have to figure out what is the story in the future when we enable native hosting APIs to specify which ALC to load the assembly into. Who should win?

**Comment by szilvaa-adsk** [comment]
@vitek-karas  I'd really like to see some solution for this in .net7. We currently use a customized ijwhost.dll to work around this problem and this will need constant synchronization to keep it up to date. Can I do anything to move this forward?

**Comment by AaronRobinsonMSFT** [comment]
> @AaronRobinsonMSFT @elinor-fung What do you think about giving managed hosts a hook into the native hosting process?

How are "managed hosts" defined in this case? The entire idea of a callback or some other mechanism bother me because of issues with loading order. How assemblies get loaded could be negatively impacted in unintentional ways. I would prefer some application centric configuration mechanism that couldn't be impacted by arbitrary assemblies.

I'm also not against simply saying C++CLI assemblies always get loaded in the default ALC and just let that be. As mentioned, the porting from .NET Framework is likely to be the primary case for C++/CLI so we could simply say this is how it is &ndash;always in the default ALC.

**Comment by batzen** [comment]
> I'm also not against simply saying C++CLI assemblies always get loaded in the default ALC and just let that be.

The same should apply to COM then and I would prefer that very much, as this is currently blocking us from migrating.

**Comment by AaronRobinsonMSFT** [comment]
@batzen Not for COM. We did change the C++/CLI behavior in .NET 7 - see https://github.com/dotnet/runtime/pull/66486. There is some nuance here as we don't want to change behavior when targeting earlier TFMs.
